### DISCUSSION TITLE
Which file is the code for USB redirection?

### DISCUSSION THREAD

**Discussion by LLeext** [FreeRDP/FreeRDP#9159] [Q&A] [answered]
Hi everyone, I am learning the USB redirection section, please ask you a question
Which file is the code for USB redirection? Channels/rdpdr?  Or Channels/urbdrc?

**Comment by akallabeth** [comment, marked as answer]
@LLeext `channels/urbdrc` is low level usb redirection, `channels/rdpdr` the device redirection (drives, smartcards and printers build upon that one)

**Reply by LLeext** [reply]
thank you

**Reply by LLeext** [reply]
I am learning USB redirection this department source code, can you provide some ways to learn this part?

**Reply by akallabeth** [reply]
[MS-RDPEUSB](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeusb/a1004d0e-99e9-4968-894b-0b924ef2f125)
